backend/msgpack/topology.go

```diff
@@ -0,0 +1,175 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"sync"
+
+	schema "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3collector/runtime"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	errNilValue                  = errors.New("nil value received")
+	errTopologyIsOpenOrClosed    = errors.New("topology is already open or closed")
+	errTopologyIsNotOpenOrClosed = errors.New("topology is not open or closed")
+	errNoEligiblePlacement       = errors.New("no eligible placement")
+)
+
+type topologyState int
+
+const (
+	topologyNotOpen topologyState = iota
+	topologyOpen
+	topologyClosed
+)
+
+// topology stores topology information.
+type topology interface {
+	// Open opens the topology.
+	Open() error
+
+	// Route routes a metric alongside its policies.
+	Route(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error
+
+	// Close closes the topology.
+	Close() error
+}
+
+type topo struct {
+	sync.RWMutex
+	runtime.Value
+
+	serverOpts    ServerOptions
+	placementOpts placementOptions
+	nowFn         clock.NowFn
+
+	state     topologyState
+	proto     *schema.PlacementSnapshots
+	placement placement
+}
+
+// newTopology creates a new topology.
+func newTopology(writerMgr instanceWriterManager, opts ServerOptions) topology {
+	clockOpts := opts.ClockOptions()
+	topology := &topo{
+		serverOpts: opts,
+		nowFn:      clockOpts.NowFn(),
+		proto:      &schema.PlacementSnapshots{},
+	}
+	topologyOpts := opts.TopologyOptions()
+	topology.placementOpts = newPlacementOptions().
+		SetClockOptions(clockOpts).
+		SetHashGenFn(topologyOpts.HashGenFn()).
+		SetInstanceWriterManager(writerMgr)
+	valueOpts := runtime.NewOptions().
+		SetInstrumentOptions(opts.InstrumentOptions()).
+		SetInitWatchTimeout(topologyOpts.InitWatchTimeout()).
+		SetKVStore(topologyOpts.KVStore()).
+		SetUnmarshalFn(topology.toPlacement).
+		SetProcessFn(topology.process)
+	topology.Value = runtime.NewValue(topologyOpts.TopologyKey(), valueOpts)
+	return topology
+}
+
+func (t *topo) Open() error {
+	t.Lock()
+	if t.state != topologyNotOpen {
+		return errTopologyIsOpenOrClosed
+	}
+	t.state = topologyOpen
+	t.Unlock()
+
+	// NB(xichen): we watch the topology updates outside the lock because
+	// otherwise the initial update will trigger the process() callback,
+	// which attempts to acquire the same lock, causing a deadlock.
+	return t.Watch()
+}
+
+func (t *topo) Route(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error {
+	t.RLock()
```

Comment:
I'm not too worried about the overhead of acquiring four RLocks. In case there's a pending writer, the RLock will be blocked immediately. In case there's no pending writers, a RLock operation is simply an atomic increment (https://golang.org/src/sync/rwmutex.go?s=894:1226#L36), and benchmark shows performing four atomic inc is fast:
```
BenchmarkAtomicIncrement-8       	50000000	        28.6 ns/op
```
Assuming we are doing 100K writes per second (this is significantly overestimating the write qps given that we are batching the timer values and only perform one write for the entire batch), each write takes 10us, so the overhead of 4 atomic incs is only 0.286%.

It is possibly to reduce rlock overhead by merging multiple RLocks into one through merging components (e.g., writer and the queue) in the write path, or removing locks from some components by making assumptions about how they interfact w/ each other, however for now I'd keep the design as is where the responsibility of each component is clean and well defined and look into the possibility of optimization when prod data/load testing shows there's need to do so.